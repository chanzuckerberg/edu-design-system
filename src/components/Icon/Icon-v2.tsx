import clsx from 'clsx';
import type { ReactNode, CSSProperties } from 'react';
import React from 'react';
import icons, { type IconName } from '../../icons/spritemap-v2';
import styles from './Icon-v2.module.css';

export type { IconName } from '../../icons/spritemap-v2';

// TODO: export union utility type of "Extract<IconName, T> | (renderProps) => ReactNode" when updating IconName usages

interface IconPropsBase {
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Either a fragment of svg elements (g, path, circle, etc.), or a single
   * svg element. Useful for creating specific icon components.
   *
   * @example
   * function CircleIcon(props: IconProps) {
   *   return (
   *     <Icon {...props}>
   *       <circle cx="50" cy="50" r="50" />
   *     </Icon>
   *   )
   * }
   */
  children?: ReactNode;
  /**
   * The SVG Color, expects a valid css color (hex, rgb, css variable, etc.).
   *
   */
  color?: string;
  /**
   * If true, the element will be displayed as a block, otherwise
   * default is inline like images
   */
  fullWidth?: boolean;
  /**
   * ID used so the svg can read the title of the SVG icon to the user when accessibility is needed
   */
  id?: string;
  /**
   * Name of icon to reference in icon sprite
   */
  name?: IconName;
  /**
   * Width/Height string (px, rem, em, vh, etc.)
   * Generally prefer using "em" as it sizes to its parent container.
   */
  size?: string;
  /**
   * viewBox for the svg, used when the svg information is passed via children.
   */
  viewBox?: string;
}

interface InformativeIconProps extends IconPropsBase {
  /**
   * The role of the icon.
   *
   * Use "informative" when the icon **_does_** provide additional meaning to other text on the
   * page. You'll be required to pass in a title to label the icon.
   */
  purpose: 'informative';
  title: string;
}

interface DecorativeIconProps extends IconPropsBase {
  /**
   * The role of the icon.
   *
   * Use "decorative" when the icon **_does not_** provide any additional context or meaning to
   * associated text. Basically the icon is for show and people don't need it to understand what's
   * on the page.
   */
  purpose: 'decorative';
}

export type IconProps = DecorativeIconProps | InformativeIconProps;

interface SvgStyle extends CSSProperties {
  '--icon-size'?: string;
}

/**
 * `import {Icon} from "@chanzuckerberg/eds";`
 *
 * Render arbitrary SVG path data while enforcing good accessibility practices.
 *
 * Icons are based on [Material Rounded](https://fonts.google.com/icons?icon.set=Material+Icons&icon.style=Rounded),
 * and are encoded in a spritemap in `src/icons`.
 */
export const Icon = (props: IconProps) => {
  const {
    children,
    className,
    color = 'currentColor',
    name,
    fullWidth = false,
    id,
    purpose,
    size,
    viewBox,
  } = props;

  const componentClassName = clsx(
    styles['icon'],
    fullWidth && styles['icon--full-width'],
    className,
  );
  const style: SvgStyle = {
    '--icon-size': size,
  };

  const svgCommonProps = {
    className: componentClassName,
    fill: color,
    height: size,
    /**
     * height/width html properties are overriden by the defaults applied in CSS module
     */
    style,
    width: size,
    xmlns: 'http://www.w3.org/2000/svg',
    viewBox: name ? icons[name].viewBox : viewBox,
  };

  // allow passing custom SVGs to render, otherwise
  // load from the spritemap of EDS icons
  const computedSvg = name ? icons[name].content : children;

  if (purpose === 'informative') {
    return (
      <svg {...svgCommonProps} role="img">
        <title id={id}>{props.title}</title>
        {computedSvg}
      </svg>
    );
  } else {
    return (
      <svg {...svgCommonProps} aria-hidden>
        {computedSvg}
      </svg>
    );
  }
};

Icon.displayName = 'Icon';
