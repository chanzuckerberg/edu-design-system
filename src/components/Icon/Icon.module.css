/* ------------------------------------*\
    #ICON
\*------------------------------------ */

/**
 * 1) Small graphic that represents functionality
 */
.icon {
  display: inline-block;

  /**
   * Size Priority:
   * 1: --icon-size, passed through props
   * 2: --icon-size-default, determined by context (text, button, pill)
   * default: 1em, same size as surrounding text
   *
   * Inspired by https://www.youtube.com/watch?v=EDyiaDJJu-4
   */
  width: var(--icon-size, var(--icon-size-default, 1em));
  height: var(--icon-size, var(--icon-size-default, 1em));
}

/**
 * Full-width icon
 * 1) A block icon fills 100% of the width of its container
 */
.icon--full-width {
  display: block;
}
