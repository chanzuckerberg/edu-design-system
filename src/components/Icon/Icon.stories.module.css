/* ------------------------------------*\
    #ICON GRID
\*------------------------------------ */

/**
 * 1) The icon grid is not for production and is meant solely to display
 *    a grid of icons within the pattern library. THerefore, it does not
 *    follow coding conventions
 */

.icon-grid {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(150px, 1fr));
  grid-gap: 1rem;
}

.icon-grid__item {
  text-align: center;
  display: box;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  flex-direction: column;
  -webkit-box-align: center;
  align-items: center;
  -webkit-box-pack: center;
  justify-content: center;
  border-radius: 4px;
  border: 1px solid #ccc;
  padding: 8px;
  height: 5.5rem;
}

.icon-grid svg {
  height: 24px;
  width: 24px;
  margin: 0 auto 1rem;
}

.icon-grid__text {
  display: block;
  font-size: 0.75rem;
}
