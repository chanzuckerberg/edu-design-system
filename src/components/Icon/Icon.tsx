import clsx from 'clsx';
import React, { ReactNode, CSSProperties } from 'react';
import { useUID } from 'react-uid';
import styles from './Icon.module.css';
import icons from '../../icons/spritemap/spritemap.svg';
import { ALL_ICONS } from '../../util/allIcons';

// generates a union type of all possible icon names
export type IconName = typeof ALL_ICONS[number];

interface IconPropsBase {
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Either a fragment of svg elements (g, path, circle, etc.), or a single
   * svg element. Useful for creating specific icon components.
   *
   * @example
   * function CircleIcon(props: IconProps) {
   *   return (
   *     <Icon {...props}>
   *       <circle cx="50" cy="50" r="50" />
   *     </Icon>
   *   )
   * }
   */
  children?: ReactNode;
  /**
   * The SVG Color, expects a valid css color (hex, rgb, etc.).
   *
   * Recommendation: if `currentColor` isn't sufficient,
   * use `EdsThemeColor` tokens from
   * `@chanzuckerberg/eds/lib/tokens/colors.ts`
   */
  color?: string;
  /**
   * If true, the element will be displayed as a block, otherwise
   * default is inline like images
   */
  fullWidth?: boolean;
  /**
   * ID used so the svg can read the title of the SVG icon to the user when accessibility is needed
   */
  id?: string;
  /**
   * Name of icon to reference in icon sprite
   */
  name?: IconName;
  /**
   * Width/Height string (px, rem, em, vh, etc.)
   * Generally prefer using "em" as it sizes to its parent container.
   */
  size?: string;
}

interface InformativeIconProps extends IconPropsBase {
  /**
   * The role of the icon.
   *
   * Use "informative" when the icon **_does_** provide additional meaning to other text on the
   * page. You'll be required to pass in a title to label the icon.
   */
  purpose: 'informative';
  title: string;
}

interface DecorativeIconProps extends IconPropsBase {
  /**
   * The role of the icon.
   *
   * Use "decorative" when the icon **_does not_** provide any additional context or meaning to
   * associated text. Basically the icon is for show and people don't need it to understand what's
   * on the page.
   */
  purpose: 'decorative';
}

export type IconProps = DecorativeIconProps | InformativeIconProps;

interface SvgStyle extends CSSProperties {
  '--icon-size'?: string;
}

/**
 * Render arbitrary SVG path data while enforcing good accessibility practices.
 *
 * If you're looking for specific icon files, look in the `src/icons` directory.
 */
export const Icon = (props: IconProps) => {
  const {
    children,
    className,
    color = 'currentColor',
    name,
    fullWidth = false,
    id,
    purpose,
    size,
  } = props;
  const generatedId = useUID();
  const titleId = id || generatedId;

  const componentClassName = clsx(
    styles['icon'],
    className,
    fullWidth && styles['icon--full-width'],
  );
  const style: SvgStyle = {
    '--icon-size': size,
  };

  const svgCommonProps = {
    className: componentClassName,
    fill: color,
    height: size,
    /**
     * height/width html properties are overriden by the defaults applied in CSS module
     */
    style,
    width: size,
    xmlns: 'http://www.w3.org/2000/svg',
  };
  // allow passing custom SVGs to render, otherwise
  // load from the spritemap of EDS icons
  const computedSvg = children || <use xlinkHref={`${icons}#${name}`} />;

  if (purpose === 'informative') {
    return (
      <svg {...svgCommonProps} role="img">
        <title id={titleId}>{props.title}</title>
        {computedSvg}
      </svg>
    );
  } else {
    return (
      <svg {...svgCommonProps} aria-hidden>
        {computedSvg}
      </svg>
    );
  }
};
