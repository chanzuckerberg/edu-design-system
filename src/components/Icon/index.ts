export { Icon as default, IconProps } from "./Icon";
export type { IconName } from "./Icon";
