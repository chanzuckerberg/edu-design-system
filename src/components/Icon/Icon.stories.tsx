import type { StoryObj, Meta } from '@storybook/react';
import kebabCase from 'lodash/kebabCase';
import React from 'react';
import { Icon, type IconProps } from './Icon';
import icons, { type IconName } from '../../icons/spritemap';
import * as ColorTokens from '../../tokens-dist/ts/colors';
import Text from '../Text';

import styles from './Icon.stories.module.css';

const meta: Meta<typeof Icon> = {
  title: 'Components/Icon',
  component: Icon,
  parameters: {
    layout: 'centered',
    badges: ['api-2.0', 'theme-2.0'],
  },
  argTypes: {
    name: {
      control: {
        type: 'select',
      },
      options: Object.keys(icons),
    },
    color: {
      control: {
        type: 'select',
      },
      // For now, take the variables and convert to equivalent tokens for the UI
      options: [
        'currentColor',
        ...Object.keys(ColorTokens)
          .filter((tokenVarName) => tokenVarName.indexOf('Icon') !== -1)
          .map((tokenVarName) => `var(--${kebabCase(tokenVarName)})`),
      ],
    },
  },
};

export default meta;
type Story = StoryObj<typeof Icon>;

export const Default: Story = {
  render: ({ name, color, ...rest }) => {
    return <Icon {...rest} color={color} name={name} />;
  },
  args: {
    name: 'close',
    purpose: 'decorative' as const,
  },
};

export const Medium: Story = {
  ...Default,
  args: {
    ...Default.args,
    size: '2em',
  },
};

export const Large: Story = {
  ...Default,
  args: {
    ...Default.args,
    size: '4em',
  },
};

/**
 * You can control the color of the icon using any valid CSS color values, including our token suite.
 *
 * If `currentColor` for the whole container isn't sufficient,
 * use a CSS variable in `color` with the token you need, or
 * style `fill` with Tailwind: https://tailwindcss.com/docs/fill
 */
export const CustomColor: Story = {
  ...Default,
  args: {
    ...Default.args,
    color: 'var(--eds-theme-color-icon-utility-critical)',
    size: '2em',
  },
};

/**
 * Icons are positioned naturally in lines of text. Use the size, color, or other props
 * to match the recommended design and layout.
 *
 * See: https://material-ui.com/components/material-icons/
 */
export const InText: Story = {
  render: (args) => {
    return (
      <Text as="p">
        The svg icon defaults to the surrounding text size (
        <Icon
          {...args}
          name="add-encircled"
          purpose="informative"
          title="icon with 1em line height"
        />
        ; 1em) by default.
      </Text>
    );
  },
};

/**
 * If your product needs icons not currently existing in the suite, you can introduce new
 * accessible icons by inserting the body of an SVG into `Icon`. Each resulting icon can be
 * treated like a standalone component, matching the recipe defined by design.
 */
export const WithChildrenSvg: Story = {
  ...Default,
  args: {
    viewBox: '0 0 24 24',
    children: (
      <path d="M11.6144 8.96051C12.2524 7.61848 13.9514 4.56304 14.905 3C15.5557 3.28671 16.1844 3.61871 16.7863 3.99342C15.9964 5.50177 14.2272 8.6643 13.5144 9.87646C14.7414 9.45266 18.4783 8.3362 20.1657 7.89646C20.4765 8.60506 20.8294 9.55291 21 10.0815C19.0439 10.5737 15.0872 11.5078 13.699 11.8109C15.335 12.2848 17.9431 13.0914 19.8642 13.7408C19.6118 14.2352 19.0345 15.2354 18.6209 15.9122C17.034 15.3311 15.0639 14.5451 13.4794 13.8934C13.9 14.3491 16.8143 17.8329 17.5178 18.7147C17.1322 19.2023 16.4217 20.0157 15.9566 20.5101C14.6058 18.7876 13.1522 16.819 12.0047 15.2149C12.0911 16.0785 12.3412 19.4119 12.43 20.8701C11.764 20.9339 10.8455 20.9841 10.093 21C10.0345 18.7739 10.0065 16.9648 10.0088 15.7253C9.39652 16.6367 7.40068 19.4734 6.97066 20.0544C6.58504 19.7263 5.80213 18.9152 5.48429 18.5985C6.41911 17.2975 8.24201 14.9939 9.14412 13.8957C8.24435 14.1965 4.57518 15.3266 3.53519 15.6046C3.35991 15.0418 3.11685 14.1076 3 13.5334C4.11945 13.2395 7.56661 12.4352 8.98987 12.1322C8.04103 11.5603 5.13607 9.68506 3.88574 8.81013C4.35316 8.17671 4.91872 7.48633 5.43521 6.88481C6.44015 7.60709 8.7772 9.32734 9.5718 9.92203C9.00857 7.9238 8.57388 5.88 8.13918 3.57646C8.9025 3.38437 9.67878 3.24503 10.4622 3.15949C10.6679 4.05949 11.4298 8.09468 11.6144 8.96051Z" />
    ),
  },
};

const IconsInGrid = (args: IconProps) => (
  <div>
    <ul className={styles['icon-grid']}>
      {(Object.keys(icons) as IconName[]).map((name) => {
        return (
          <li className={styles['icon-grid__item']} key={name}>
            <Icon className={styles['icon-grid__icon']} name={name} {...args} />
            <span className={styles['icon-grid__text']}>{name}</span>
          </li>
        );
      })}
    </ul>
  </div>
);

/**
 * Grid of all the available icons. Use the controls to change color, or other attributes
 *
 * **NOTE**: some icons marked as deprecated will be removed in future releases.
 */
export const IconGrid: Story = {
  render: (args) => <IconsInGrid {...args} />,
  parameters: {
    layout: 'padded',
  },
};
