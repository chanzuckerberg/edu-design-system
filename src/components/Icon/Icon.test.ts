import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as IconStoryFile from './Icon.stories';

describe('<Icon />', () => {
  generateSnapshots(IconStoryFile);
});
