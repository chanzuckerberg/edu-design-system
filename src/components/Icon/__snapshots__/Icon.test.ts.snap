// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> CustomColor story renders snapshot 1`] = `
<svg
  aria-hidden="true"
  class="icon"
  fill="#A6A3EE"
  height="2em"
  style="--icon-size: 2em;"
  viewBox="0 0 24 24"
  width="2em"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6.4 19L5 17.6L10.6 12L5 6.4L6.4 5L12 10.6L17.6 5L19 6.4L13.4 12L19 17.6L17.6 19L12 13.4L6.4 19Z"
  />
</svg>
`;

exports[`<Icon /> Default story renders snapshot 1`] = `
<svg
  aria-hidden="true"
  class="icon"
  fill="currentColor"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6.4 19L5 17.6L10.6 12L5 6.4L6.4 5L12 10.6L17.6 5L19 6.4L13.4 12L19 17.6L17.6 19L12 13.4L6.4 19Z"
  />
</svg>
`;

exports[`<Icon /> FullScreen story renders snapshot 1`] = `
<svg
  aria-hidden="true"
  class="icon"
  fill="currentColor"
  height="100vh"
  style="--icon-size: 100vh;"
  viewBox="0 0 24 24"
  width="100vh"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6.4 19L5 17.6L10.6 12L5 6.4L6.4 5L12 10.6L17.6 5L19 6.4L13.4 12L19 17.6L17.6 19L12 13.4L6.4 19Z"
  />
</svg>
`;

exports[`<Icon /> InText story renders snapshot 1`] = `
<p
  class="text text--body"
>
  The svg icon defaults to the surrounding text size (
  <svg
    class="icon"
    fill="currentColor"
    role="img"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title
      id="6"
    >
      icon with 1em line height
    </title>
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
    />
  </svg>
  ; 1em), but often looks better with the line height (
  <svg
    class="icon"
    fill="currentColor"
    height="2em"
    role="img"
    style="--icon-size: 2em;"
    viewBox="0 0 24 24"
    width="2em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title
      id="7"
    >
      icon with 2em line height
    </title>
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
    />
  </svg>
  ; 2em) which is harder to determine. Take a look at the icons available in
   
  <a
    href="https://material-ui.com/components/material-icons/"
    rel="noreferrer"
    target="_blank"
  >
    https://material-ui.com/components/material-icons/
  </a>
  , currently we only support the filled icons.
</p>
`;

exports[`<Icon /> Large story renders snapshot 1`] = `
<svg
  aria-hidden="true"
  class="icon"
  fill="currentColor"
  height="4em"
  style="--icon-size: 4em;"
  viewBox="0 0 24 24"
  width="4em"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6.4 19L5 17.6L10.6 12L5 6.4L6.4 5L12 10.6L17.6 5L19 6.4L13.4 12L19 17.6L17.6 19L12 13.4L6.4 19Z"
  />
</svg>
`;

exports[`<Icon /> Medium story renders snapshot 1`] = `
<svg
  aria-hidden="true"
  class="icon"
  fill="currentColor"
  height="2em"
  style="--icon-size: 2em;"
  viewBox="0 0 24 24"
  width="2em"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6.4 19L5 17.6L10.6 12L5 6.4L6.4 5L12 10.6L17.6 5L19 6.4L13.4 12L19 17.6L17.6 19L12 13.4L6.4 19Z"
  />
</svg>
`;

exports[`<Icon /> WithChildrenSvg story renders snapshot 1`] = `
<svg
  aria-hidden="true"
  class="icon"
  fill="currentColor"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M11.6144 8.96051C12.2524 7.61848 13.9514 4.56304 14.905 3C15.5557 3.28671 16.1844 3.61871 16.7863 3.99342C15.9964 5.50177 14.2272 8.6643 13.5144 9.87646C14.7414 9.45266 18.4783 8.3362 20.1657 7.89646C20.4765 8.60506 20.8294 9.55291 21 10.0815C19.0439 10.5737 15.0872 11.5078 13.699 11.8109C15.335 12.2848 17.9431 13.0914 19.8642 13.7408C19.6118 14.2352 19.0345 15.2354 18.6209 15.9122C17.034 15.3311 15.0639 14.5451 13.4794 13.8934C13.9 14.3491 16.8143 17.8329 17.5178 18.7147C17.1322 19.2023 16.4217 20.0157 15.9566 20.5101C14.6058 18.7876 13.1522 16.819 12.0047 15.2149C12.0911 16.0785 12.3412 19.4119 12.43 20.8701C11.764 20.9339 10.8455 20.9841 10.093 21C10.0345 18.7739 10.0065 16.9648 10.0088 15.7253C9.39652 16.6367 7.40068 19.4734 6.97066 20.0544C6.58504 19.7263 5.80213 18.9152 5.48429 18.5985C6.41911 17.2975 8.24201 14.9939 9.14412 13.8957C8.24435 14.1965 4.57518 15.3266 3.53519 15.6046C3.35991 15.0418 3.11685 14.1076 3 13.5334C4.11945 13.2395 7.56661 12.4352 8.98987 12.1322C8.04103 11.5603 5.13607 9.68506 3.88574 8.81013C4.35316 8.17671 4.91872 7.48633 5.43521 6.88481C6.44015 7.60709 8.7772 9.32734 9.5718 9.92203C9.00857 7.9238 8.57388 5.88 8.13918 3.57646C8.9025 3.38437 9.67878 3.24503 10.4622 3.15949C10.6679 4.05949 11.4298 8.09468 11.6144 8.96051Z"
  />
</svg>
`;
